src/main/java/com/google/devtools/build/lib/vfs/FileSystemUtils.java

```diff
@@ -374,6 +377,18 @@ public static ByteSink asByteSink(final Path path) {
    */
   @ThreadSafe  // but not atomic
   public static void copyFile(Path from, Path to) throws IOException {
+    var fromNio = from.getFileSystem().getNioPath(from.asFragment());
+    var toNio = to.getFileSystem().getNioPath(to.asFragment());
+    if (fromNio != null && toNio != null) {
+      // Fast path: Files.copy uses various optimizations such as kernel buffers (sendfile on Unix)
+      // or copy-on-write (clonefile on macOS, copy_file_range on Linux with a supported file system).
+      try {
+        java.nio.file.Files.copy(fromNio, toNio, REPLACE_EXISTING, COPY_ATTRIBUTES);
+      } catch (IOException e) {
+        throw translateNioToIoException(from.asFragment(), e);
+      }
+      return;
+    }
```

Comment:
Can you also replace `ByteStreams.copy(in, out)` with `in.transferTo(out)` below? It can still do in-kernel I/O when both sides are files, and falls back to essentially equivalent code otherwise.

Reply:
I did that and realized that the copy fallback in `moveFile` can also be optimized in this way.